packages/yew-router/src/service.rs

```diff
@@ -1,90 +1,160 @@
-use crate::utils::base_url;
-use crate::Routable;
-use gloo::events::EventListener;
-use serde::{Deserialize, Serialize};
-use wasm_bindgen::JsValue;
-use web_sys::Event;
+use std::cell::RefCell;
+use std::rc::Rc;
+
+use anymap::AnyMap;
 use yew::Callback;
 
-/// Navigate to a specific route.
-pub fn push_route(route: impl Routable) {
-    push_impl(route.to_path())
-}
+use super::history::{self, HistoryListener, Route};
+use crate::Routable;
 
-/// Navigate to a specific route with query parameters.
-///
-/// This should be used in cases where [`Link`](crate::prelude::Link) is insufficient.
-pub fn push_route_with_query<S>(
-    route: impl Routable,
-    query: S,
-) -> Result<(), serde_urlencoded::ser::Error>
-where
-    S: Serialize,
-{
-    let mut url = route.to_path();
-    let query = serde_urlencoded::to_string(query)?;
-    if !query.is_empty() {
-        url.push_str(&format!("?{}", query));
-    }
+type Entry<'a, T> = anymap::Entry<'a, dyn anymap::any::Any, T>;
 
-    push_impl(url);
+#[derive(Debug, Clone)]
+#[non_exhaustive]
+enum RouterAction<T: Routable> {
+    Push(T),
+    Replace(T),
+}
+
+/// Specializes the history API for a particular `Routable` type
+pub(crate) struct RouterState<T: Routable> {
+    last_route: Rc<T>,
+    subscribers: Vec<Callback<Rc<T>>>,
+    _listener: HistoryListener,
+}
 
-    Ok(())
+thread_local! {
+    static ROUTER_STATE: RefCell<AnyMap> = RefCell::new(AnyMap::new());
 }
 
-fn push_impl(url: String) {
-    let history = yew::utils::window().history().expect("no history");
-    let base = base_url();
-    let path = match base {
-        Some(base) => {
-            let path = format!("{}{}", base, url);
-            if path.is_empty() {
-                "/".to_string()
+impl<T: Routable> RouterState<T> {
+    fn with<R>(f: impl FnOnce(Entry<Self>) -> R) -> R {
+        ROUTER_STATE.with(|state| f(state.borrow_mut().entry()))
+    }
+    fn try_with<R>(mut f: impl FnMut(Entry<Self>) -> Result<R, T>) -> R {
+        match Self::with(&mut f) {
+            Ok(res) => return res,
+            Err(error) => Self::handle_error(error),
+        };
+        match Self::with(&mut f) {
+            Ok(res) => res,
+            Err(_) => panic!("Bug in `Routable` implementation"),
+        }
```

Comment:
Why call the function twice?

Reply:
The first call may result in a redirect if the current URL does not match. The second call runs after the redirect, and so should never return a non-matching route.

Reply:
I feel like a better way to do this would be remove this function entirely and only add the second call where exactly it is needed, with a comment explaining why it's there.

Reply:
I'm not sure why you'd intentionally create code duplication like that?

Reply:
I assumed that only one of call sites required the 2nd call/match so I thought it would be better to put that there. If that isn't the case and both (iirc) of the call sites require it, I think a comment here explaining why 2 calls are used would suffice

Reply:
Added